### ISSUE TITLE
Error in mediaTrackPlay sequene

### ISSUE THREAD

**Issue by dipan-majumder** [Adobe-Marketing-Cloud/video-heartbeat-v2#19]
Hi,

So i am not firing `mediaTrackPlay` after `mediaTrackLoad` event in the beginning whenever there is preroll, because ad is playing and not the content, instead I am firing `mediaTrackEvent` with `MEDIA_AD_BREAK_START` constant after `mediaTrackLoad`.

But when I do this, it throws type mismatch error in below line-  
>_adb_logger().debug("Timer -  " + m._name + " timer already stopped.")

because the timer is not initialized and ``m._name`` is invalid.

Any suggestions? 
Let me know if I am clear enough explaining the issue or you need more info.

Thanks
Dipan.

**Comment by praveek** [comment]
@dipan-majumder Are you calling ```mediaTrackStart``` after ```mediaTrackLoad``` ? ```mediaTrackStart``` should be called when the user initiates the video playback to denote start of playback.

**Comment by praveek** [comment]
Yes. We except ```mediaTrackPlay``` to be called after ```MEDIA_AD_START``` to denote the preroll ad has started playing.  If you still see the error after calling ```mediaTrackPlay``` can you open a ticket with the client care team with debug logs attached for us to investigate further.

**Comment by dipan-majumder** [comment]
Do I need to call `mediaTrackPlay` after every `mediaTrackEvent` call like after every `MEDIA_AD_START`, `MEDIA_AD_COMPLETE` etc?

**Comment by praveek** [comment]
You have to call ```mediaTrackPlay``` only after ```MEDIA_AD_START``` to denote start of ad playback. If the content does not have any preroll ad, call ```mediaTrackPlay``` after ```mediaTrackStart``` to denote start of content. 

You can contact client care team using this url. https://helpx.adobe.com/marketing-cloud/contact-support.html

**Comment by dipan-majumder** [comment]
I am getting `stall` event for Live stream. what might be the reason? and what is meaning of this event? Do I have to handle live stream separately than VOD?

**Comment by himanshuphalak** [comment]
Hi @dipan-majumder, Stall tracking is a feature introduced in 2.x SDK versions where if playhead value doesn't progress every second, VA tracking gets into stall state and will start sending out events for the same. To fix this issue please make sure you are incrementing playhead value every second as playback occurs using **mediaUpdatePlayhead** API. 
For VOD stream we recommend content time to be reported every second and for LIVE streams we recommend UTC time to be reported every second.

**Comment by dipan-majumder** [comment]
Thanks a lot @himanshuphalak

**Comment by stevenjhan** [comment]
Hi @praveek,
I am working with @dipan-majumder on the Roku implementation. 

At the start of the video, I instructed @dipan-majumder to call mediaTrackPlay() after mediaTrackEvent(ADBMobile().MEDIA_AD_START,adinfo) if there is a pre-roll, and after mediaTrackEvent(ADBMobile().MEDIA_CHAPTER_START,chapterinfo) if there is not. In both of these scenarios, the mediaTrackEvent(...) API method is called after mediaTrackLoad(mediainfo,contextdata) and mediaTrackStart() are called. 

That said, @dipan-majumder mentioned that if he doesn't call mediaTrackPlay() after mediaTrackStart(), the Adobe Roku SDK will throw an error. From a reporting perspective, this cannot happen, or else content starts will increment too soon. In the case of an asset with a pre-roll ad, this can inflate content starts from 10-15% given historical user drop off from video initiate --> pre-roll ad --> content start. The Roku SDK, in the absence of any track..() calls should trigger HB stall events in lieu of the SDK throwing an error. 

Can you please provide help with what's the best approach here? Overinflating content starts is not an option for us.

**Comment by praveek** [comment]
Hi @stevenjhan 

Calling mediaTrackPlay() after mediaTrackEvent(ADBMobile().MEDIA_AD_START, adInfo) will not inflate the content start. 

That said, the crash here happens particularly when calling mediaTrackPause(), mediaTrackEvent(ADBMobile(). MEDIA_SEEK_START) or mediaTrackEvent(ADBMobile(). MEDIA_BUFFER_START) before calling mediaTrackPlay() due to a code path which tries to log uninitialized timer variable.
Commenting the following line should fix the issue.
```_adb_logger().debug("Timer - " + m._name + " timer already stopped.") ```

We have fixed these issues in version 2.1 together with lot of enhancements and will be releasing the new SDK in the first week of August.

Thanks,
Praveen

**Comment by dipan-majumder** [comment]
Hi,

So we are calling the SDK for an asset with pre-roll in below sequence:

mediaTrackLoad()
mediaTrackStart()
mediaTrackEvent(m.adbmobile.MEDIA_AD_BREAK_START, adbreakinfo, adPodContextdata)
mediaTrackPlay()
mediaTrackEvent(m.adbmobile.MEDIA_AD_START, adinfo, adContextdata)
mediaTrackPlay()
…..and so on.

But the heartbeat pings we see, have an extra “play-main” event even before the ad start ping. See the sequence we get below:

s:event:type=start, s:asset:type=main
s:event:type=aa_start, s:asset:type=main
s:event:type=buffer, s:asset:type=main
**s:event:type=play, s:asset:type=main**
s:event:type=start, s:asset:type=ad
s:event:type=aa_ad_start, s:asset:type=ad
s:event:type=play, s:asset:type=ad

From a reporting perspective, this cannot happen, or else content starts will increment too soon and can inflate content starts. 
Even for assets without preroll, we see this call before the chapter start event.
Can you please help us to get rid of this errant call?
Thanks!

**Comment by himanshuphalak** [comment]
@dipan-majumder, I see you are calling trackPlay between AdBreakStart and AdStart trackEvent calls. Curious to know if there's any specific reason for doing this? 
Since AdBreakStart only sets pod information internally for the SDK, it is not considered as an Ad STATE. SDK will get into AD STATE only when AdStart API is executed. Calling mediaTrackPlay API anytime before AdStart will cause main:play to occur.

**Comment by dipan-majumder** [comment]
ok, I will make the changes you suggested and check, thanks!

**Comment by dipan-majumder** [comment]
@himanshuphalak, its also happening for assets without pre-rolls - 

So we are calling the SDK for an asset without pre-roll in below sequence:

mediaTrackLoad()
mediaTrackStart()
mediaTrackEvent(m.adbmobile.MEDIA_CHAPTER_START, chapterinfo, chapter_contextdata)
mediaTrackPlay()

But the heartbeat pings we see, have an extra “play-main” event even before the chapter start ping. See the sequence we get below:

s:event:type=start, s:asset:type=main
s:event:type=aa_start, s:asset:type=main
**s:event:type=play, s:asset:type=main**
s:event:type=buffer, s:asset:type=main
s:event:type=chapter_start, s:asset:type=main

Any suggestions for assets without preroll?

**Comment by himanshuphalak** [comment]
@dipan-majumder Chapter tracking is a specialized form of main content tracking so "Content Start" metric should not be impacted if there is a chapter at the beginning or not. Hence, whenever where is a **trackChapterStart** and **trackPlay** API calls together, SDK does not control the order of events going out of the SDK and **chapter_start** may not be guaranteed to go out before **play**.

**Comment by dipan-majumder** [comment]
@himanshuphalak, 
I am not calling **trackPlay** after pod start, i call play only after ad start and chapter start. But still very randomly I get the **play main** event call sometimes before ad start (when with preroll) and chapter start (when without preroll). 
Is **trackplay** event running in a different thread and you cannot always guarantee the order that it will be called only after ad start (when with preroll) and chapter start (when without preroll)?
Also is there anyway we can guarantee that **play-main** will be called only after ad start (when with preroll) and chapter start (when without preroll)?

**Comment by dipan-majumder** [comment]
@himanshuphalak / @praveek 
Please let me know if its an issue with the adobe sdk, if not how do we fix it?

**Comment by praveek** [comment]
@dipan-majumder Can you verify this with the latest 2.1 release (https://github.com/Adobe-Marketing-Cloud/video-heartbeat-v2/releases/tag/roku-v2.1.0)

The root cause **"Timer: type mismatch error"** has been fixed and you don't have to call mediaTrackPlay till the content / ad starts playing. 

Seeing **"main:play"** when calling **chapter_start** can happen as chapter tracking is a specialized form of content tracking and roku might send the pings out of order. 

You should not see **"main:play"** events when calling **mediaTrackPlay** after **mediaTrackEvent(m.adbmobile.MEDIA_AD_START)** . 

If you still see any issues, please share us debug and charles logs for us to investigate further. 

Thanks,
Praveen

**Comment by dipan-majumder** [comment]
Hi,
We still see **"main:play"** randomly before ad start incase of preroll in the new sdk as well. 

We also have a new issue - sometimes we get **"main:start"** twice like in below sequence:

main:start
main:aa_start
main-start

And below is sequence we fire them:
mediaTrackSessionStart()
mediaTrackEvent(MEDIA_AD_BREAK_START)
mediaTrackEvent(MEDIA_AD_START)
mediaTrackPlay()

Thanks!

**Comment by praveek** [comment]
Seeing multiple **"main:start"** is not an issue. The accumulated duration from the **"main:start"** calls is credited as timeToStart in the analytics dashboard.

Regarding the **"main:play"**, can you enable logging and check whether the call to mediaTrackEvent(MEDIA_AD_START) throws any error. It will be helpful if we can get debug logs to look more into the issue.

**Comment by praveek** [comment]
@dipan-majumder Do you have any updates on this issue ?

**Comment by dipan-majumder** [comment]
@praveek we are still verifying all scenarios, I see it happening very very randomly, will keep you posted, thanks!
